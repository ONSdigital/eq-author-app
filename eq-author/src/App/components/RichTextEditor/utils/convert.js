import { convertToHTML, convertFromHTML } from "draft-convert";
import { EditorState } from "draft-js";

export const toHTML = entityMap => {
  const entityToHTML = (entity, originalText) => {
    const mapper = entityMap[entity.type];
    return mapper ? mapper(entity) : originalText;
  };

  const convert = convertToHTML({ entityToHTML });

  return editorState => convert(editorState.getCurrentContent());
};

export const fromHTML = htmlToEntity => {
  const convert = convertFromHTML({ htmlToEntity });

  return (html, decorator) =>
    EditorState.createWithContent(convert(html), decorator);
};
