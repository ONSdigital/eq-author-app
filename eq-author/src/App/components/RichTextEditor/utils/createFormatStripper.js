import { mapKeys } from "lodash";
import { filterEditorState } from "draftjs-filters";
import { filterConfig } from "App/components/RichTextEditor/entities/PipedValue";

const mapper = {
  bold: { format: "BOLD", type: "styles" },
  emphasis: { format: "ITALIC", type: "styles" },
  list: { format: "unordered-list-item", type: "blocks" },
  heading: { format: "header-two", type: "blocks" }
};

export default function createFormatStripper(controls) {
  const filterConfiguration = {
    blocks: [],
    styles: [],
    entities: [filterConfig],
    maxNesting: 0,
    whitespacedCharacters: []
  };

  mapKeys(controls, (value, key) => {
    if (mapper[key] && controls[key]) {
      filterConfiguration[mapper[key].type].push(mapper[key].format);
    }
  });

  return editorState => filterEditorState(filterConfiguration, editorState);
}
