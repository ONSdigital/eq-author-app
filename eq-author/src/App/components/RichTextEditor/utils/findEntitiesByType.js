export default entityType => {
  return function findEntitiesByType(contentBlock, callback, contentState) {
    contentBlock.findEntityRanges(character => {
      const entityKey = character.getEntity();
      return (
        character.getEntity() !== null &&
        contentState.getEntity(entityKey).getType() === entityType
      );
    }, callback);
  };
};
