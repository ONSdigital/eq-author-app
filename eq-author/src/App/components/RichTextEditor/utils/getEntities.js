function getEntities(contentState, entityType) {
  const entities = [];

  contentState.getBlocksAsArray().forEach(block => {
    let selectedEntity = null;

    block.findEntityRanges(
      character => {
        const entityKey = character.getEntity();

        if (entityKey === null) {
          return false;
        }

        const entity = contentState.getEntity(entityKey);

        if (entityType && entity.getType() !== entityType) {
          return false;
        }

        selectedEntity = {
          entityKey,
          entity,
          blockKey: block.getKey()
        };

        return true;
      },
      (start, end) => {
        entities.push({ ...selectedEntity, start, end });
      }
    );
  });

  return entities;
}

export default getEntities;
