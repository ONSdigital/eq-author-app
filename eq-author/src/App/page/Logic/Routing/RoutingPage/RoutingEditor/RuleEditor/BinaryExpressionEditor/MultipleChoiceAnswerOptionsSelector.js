import React from "react";
import styled from "styled-components";
import { includes, get } from "lodash";
import { PropTypes } from "prop-types";
import { TransitionGroup } from "react-transition-group";

import CheckboxChipTransition from "./CheckboxChipTransition";
import { colors } from "constants/theme";
import { RADIO } from "constants/answer-types";
import { Select } from "components/Forms";

import TextButton from "components/buttons/TextButton";
import ToggleChip from "components/buttons/ToggleChip";
import Popover from "./CheckboxSelectorPopup";
import CheckboxOptionPicker from "./CheckboxOptionPicker";
import CheckboxChip from "./CheckboxChip";

const answerConditions = {
  UNANSWERED: "Unanswered",
  ALLOF: "AllOf",
  ANYOF: "AnyOf",
};

const MultipleChoiceAnswerOptions = styled.div`
  align-items: center;
  padding: 1em 0;
  display: inline-flex;
  flex-flow: row wrap;
`;

const Label = styled.label`
  font-size: 1em;
  font-weight: bold;
  color: ${colors.textLight};
`;

const ConditionSelect = styled(Select)`
  display: inline-block;
  width: auto;
  margin: 0 0.5em;
  padding: 0.3em 2em 0.3em 0.5em;
`;

const ChooseButton = styled(TextButton)`
  margin: 0.25rem;
`;

const SelectedOptions = styled.div`
  display: flex;
  align-items: center;
  flex-wrap: wrap;
`;

class MultipleChoiceAnswerOptionsSelector extends React.Component {
  constructor(props) {
    super(props);
    this.state = { showPopup: false };
  }

  static propTypes = {
    expression: PropTypes.shape({
      left: PropTypes.shape({
        options: PropTypes.arrayOf(
          PropTypes.shape({
            id: PropTypes.string.isRequired,
            label: PropTypes.string,
          })
        ),
      }).isRequired,
      right: PropTypes.shape({
        options: PropTypes.arrayOf(
          PropTypes.shape({
            id: PropTypes.string.isRequired,
          })
        ),
      }),
    }).isRequired,
    onRightChange: PropTypes.func.isRequired,
    onConditionChange: PropTypes.func.isRequired,
  };

  get selectedOptionIds() {
    return get(this.props.expression.right, "options", []).map(({ id }) => id);
  }

  handleRadioChange = ({ name: id, value: checked }) => {
    let newSelectedOptions;
    if (checked) {
      newSelectedOptions = [...this.selectedOptionIds, id];
    } else {
      newSelectedOptions = this.selectedOptionIds.filter(
        selectedId => selectedId !== id
      );
    }
    this.props.onRightChange({ selectedOptions: newSelectedOptions });
  };

  handleCheckboxUnselect = id => {
    const { expression } = this.props;
    const selectedOptions = expression.right.options
      .filter(option => option.id !== id)
      .map(({ id }) => id);
    this.props.onRightChange({ selectedOptions });
  };

  handleConditionChange = ({ value }) => {
    this.props.onConditionChange(value);
  };

  handlePickerClose = () => {
    this.setState({ showPopup: false });
  };

  renderRadioOptionSelector() {
    const { expression } = this.props;
    const options = get(expression, "left.options", []);

    return (
      <MultipleChoiceAnswerOptions data-test="options-selector">
        {options.map(option => (
          <ToggleChip
            key={option.id}
            name={option.id}
            title={option.label}
            checked={includes(this.selectedOptionIds, option.id)}
            onChange={this.handleRadioChange}
          >
            {option.label || <strong>Unlabelled option</strong>}
          </ToggleChip>
        ))}
      </MultipleChoiceAnswerOptions>
    );
  }

  renderCheckboxOptionSelector() {
    const { expression } = this.props;
    return (
      <MultipleChoiceAnswerOptions data-test="options-selector">
        <Label>Match</Label>
        <ConditionSelect
          onChange={this.handleConditionChange}
          defaultValue={expression.condition}
          data-test="condition-dropdown"
        >
          <option value={answerConditions.ANYOF}>Any of</option>
          <option value={answerConditions.ALLOF}>All of</option>
          <option value={answerConditions.UNANSWERED}>Unanswered</option>
        </ConditionSelect>
        {expression.condition !== answerConditions.UNANSWERED && (
          <>
            <TransitionGroup component={SelectedOptions}>
              {get(expression, "right.options", []).map(option => (
                <CheckboxChipTransition key={option.id}>
                  <CheckboxChip
                    key={option.id}
                    id={option.id}
                    onRemove={this.handleCheckboxUnselect}
                  >
                    {option.label}
                  </CheckboxChip>
                </CheckboxChipTransition>
              ))}
            </TransitionGroup>
            <ChooseButton
              onClick={() => {
                this.setState({
                  showPopup: true,
                });
              }}
            >
              Choose
            </ChooseButton>
          </>
        )}
        {this.state.showPopup && (
          <Popover isOpen onClose={this.handlePickerClose}>
            <CheckboxOptionPicker
              expression={expression}
              onClose={this.handlePickerClose}
              onRightChange={this.props.onRightChange}
            />
          </Popover>
        )}
      </MultipleChoiceAnswerOptions>
    );
  }

  render() {
    const { expression } = this.props;
    const answerType = get(expression, "left.type");
    if (answerType === RADIO) {
      return this.renderRadioOptionSelector();
    } else {
      return this.renderCheckboxOptionSelector();
    }
  }
}

export default MultipleChoiceAnswerOptionsSelector;
