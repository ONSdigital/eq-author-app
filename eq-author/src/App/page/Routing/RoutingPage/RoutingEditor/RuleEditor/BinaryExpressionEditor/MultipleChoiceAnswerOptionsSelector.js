import React from "react";
import styled from "styled-components";

import ToggleChip from "components/buttons/ToggleChip";
import { includes, get } from "lodash";

import { PropTypes } from "prop-types";

const MultipleChoiceAnswerOptions = styled.div`
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  padding: 1em 0;
`;

class MultipleChoiceAnswerOptionsSelector extends React.Component {
  static propTypes = {
    expression: PropTypes.shape({
      left: PropTypes.shape({
        options: PropTypes.arrayOf(
          PropTypes.shape({
            id: PropTypes.string.isRequired,
            label: PropTypes.string,
          })
        ),
      }).isRequired,
      right: PropTypes.shape({
        options: PropTypes.arrayOf(
          PropTypes.shape({
            id: PropTypes.string.isRequired,
          })
        ),
      }),
    }).isRequired,
    onRightChange: PropTypes.func.isRequired,
  };

  get selectedOptionIds() {
    return get(this.props.expression.right, "options", []).map(({ id }) => id);
  }

  handleChange = ({ name: id, value: checked }) => {
    let newSelectedOptions;
    if (checked) {
      newSelectedOptions = [...this.selectedOptionIds, id];
    } else {
      newSelectedOptions = this.selectedOptionIds.filter(
        selectedId => selectedId !== id
      );
    }
    this.props.onRightChange({ selectedOptions: newSelectedOptions });
  };

  render() {
    const { expression } = this.props;
    const options = get(expression, "left.options", []);

    return (
      <MultipleChoiceAnswerOptions data-test="options-selector">
        {options.map(option => (
          <ToggleChip
            key={option.id}
            name={option.id}
            title={option.label}
            checked={includes(this.selectedOptionIds, option.id)}
            onChange={this.handleChange}
          >
            {option.label || <strong>Unlabelled option</strong>}
          </ToggleChip>
        ))}
      </MultipleChoiceAnswerOptions>
    );
  }
}

export default MultipleChoiceAnswerOptionsSelector;
