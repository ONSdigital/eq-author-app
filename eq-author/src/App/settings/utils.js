export const getThemeSettingsErrorCount = (questionnaire) =>
  questionnaire?.themeSettings?.validationErrorInfo?.totalCount +
    questionnaire?.validationErrorInfo?.totalCount || 0;
