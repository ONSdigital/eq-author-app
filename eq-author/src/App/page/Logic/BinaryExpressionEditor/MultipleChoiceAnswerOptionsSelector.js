import React from "react";
import styled from "styled-components";
import { includes, get, find } from "lodash";
import { PropTypes, string } from "prop-types";
import { TransitionGroup } from "react-transition-group";

import CheckboxChip from "./CheckboxChip";
import CheckboxChipTransition from "./CheckboxChipTransition";
import CheckboxOptionPicker from "./CheckboxOptionPicker";
import Popover from "./CheckboxSelectorPopup";
import ValidationError from "components/ValidationError";

import {
  rightSideErrors,
  OPERATOR_REQUIRED,
} from "constants/validationMessages";
import { colors } from "constants/theme";
import { RADIO } from "constants/answer-types";
import { Select } from "components/Forms";

import TextButton from "components/buttons/TextButton";
import ToggleChip from "components/buttons/ToggleChip";
import SecondaryConditionSelector from "./SecondaryConditionSelector";
import { enableOn } from "utils/featureFlags";

const answerConditions = {
  UNANSWERED: "Unanswered",
  COUNTOF: "CountOf",
  ALLOF: "AllOf",
  ANYOF: "AnyOf",
  ONEOF: "OneOf",
  NOTANYOF: "NotAnyOf",
};

const MultipleChoiceAnswerOptions = styled.div`
  align-items: center;
  display: inline-flex;
  flex-flow: row wrap;
  width: 100%;
  margin-top: 0.75em;
  margin-bottom: 1em;
  padding: 0.5em;
  border-radius: 4px;
  border: 2px solid ${colors.lighterGrey};
  ${({ hasError }) =>
    hasError &&
    `
    border-color: ${colors.errorPrimary};
    outline-color: ${colors.errorPrimary};
    box-shadow: 0 0 0 2px ${colors.errorPrimary};
    border-radius: 4px;
    margin-bottom: 0.5em;
  `}
`;

const Label = styled.label`
  font-size: 1em;
  font-weight: bold;
  color: ${colors.textLight};
`;

const ConditionSelect = styled(Select)`
  display: inline-block;
  width: auto;
  margin: 0 0.5em;
  padding: 0.3em 2em 0.3em 0.5em;
  ${({ hasError }) => hasError && `border: thin solid #d0021b`}
`;

const ChooseButton = styled(TextButton)`
  font-size: 1rem;
  background: ${colors.white};
  padding: 0.3em 0.5em;
  border-radius: 4px;
  color: ${colors.primary};
  border: 1px solid ${colors.primary};
  display: inline-flex;
  align-items: center;
  margin: 0.25rem;
  max-width: 12em;
  letter-spacing: 0;

  &:hover {
    color: ${colors.white};
    background: ${colors.primary};
  }

  &:focus,
  &:active {
    outline-width: 0;
  }

  &:focus {
    box-shadow: 0 0 0 3px ${colors.tertiary};
    outline: none;
  }
`;

const SelectedOptions = styled.div`
  display: flex;
  align-items: center;
  flex-wrap: wrap;
`;

class MultipleChoiceAnswerOptionsSelector extends React.Component {
  constructor(props) {
    super(props);
    this.state = { showPopup: false };
  }

  static propTypes = {
    expression: PropTypes.shape({
      condition: string,
      left: PropTypes.shape({
        options: PropTypes.arrayOf(
          PropTypes.shape({
            id: PropTypes.string.isRequired,
            label: PropTypes.string,
          })
        ),
      }).isRequired,
      right: PropTypes.shape({
        options: PropTypes.arrayOf(
          PropTypes.shape({
            id: PropTypes.string.isRequired,
          })
        ),
      }),
    }).isRequired,
    onRightChange: PropTypes.func.isRequired,
    onConditionChange: PropTypes.func.isRequired,
    groupErrorMessage: PropTypes.string,
  };

  get selectedOptionIds() {
    return get(this.props.expression.right, "options", []).map(({ id }) => id);
  }

  handleRadioChange = ({ name: id, value: checked }) => {
    let newSelectedOptions;
    if (checked) {
      newSelectedOptions = [...this.selectedOptionIds, id];
    } else {
      newSelectedOptions = this.selectedOptionIds.filter(
        (selectedId) => selectedId !== id
      );
    }
    this.props.onRightChange({ selectedOptions: newSelectedOptions });
  };

  handleCheckboxUnselect = (id) => {
    const { expression } = this.props;
    const selectedOptions = expression.right.options
      .filter((option) => option.id !== id)
      .map(({ id }) => id);
    this.props.onRightChange({ selectedOptions });
  };

  handleConditionChange = ({ value }) => {
    this.props.onConditionChange(value);
  };

  handlePickerClose = () => {
    this.setState({ showPopup: false });
  };

  handleError = () => {
    const { expression, groupErrorMessage } = this.props;
    let message = groupErrorMessage;

    const error = find(
      expression.validationErrorInfo.errors,
      (error) =>
        error.errorCode.includes("ERR_RIGHTSIDE") ||
        error.errorCode.includes(
          "ERR_GROUP_MIXING_EXPRESSIONS_WITH_OR_STND_OPTIONS_IN_AND"
        )
    );

    if (
      expression.validationErrorInfo.errors.some(
        ({ field }) => field === "secondaryCondition"
      )
    ) {
      message = OPERATOR_REQUIRED;
    } else if (error) {
      if (expression.condition === "CountOf") {
        message = rightSideErrors[error.errorCode].message;
      } else {
        message =
          message ||
          rightSideErrors[error.errorCode].optionsMessage ||
          rightSideErrors[error.errorCode].message;
      }
    }

    return message ? <ValidationError>{message}</ValidationError> : null;
  };

  renderRadioOptionSelector(hasError) {
    const { expression } = this.props;
    const options = get(expression, "left.options", []);

    return (
      <>
        <MultipleChoiceAnswerOptions
          data-test="options-selector"
          hasError={hasError}
        >
          <Label>Match</Label>
          <ConditionSelect
            onChange={this.handleConditionChange}
            defaultValue={expression.condition}
            data-test="condition-dropdown"
          >
            <option value={answerConditions.ONEOF}>One of</option>
            <option value={answerConditions.UNANSWERED}>Unanswered</option>
          </ConditionSelect>
          {expression.condition !== answerConditions.UNANSWERED &&
            options.map((option) => (
              <ToggleChip
                key={option.id}
                name={option.id}
                title={option.label}
                checked={includes(this.selectedOptionIds, option.id)}
                onChange={this.handleRadioChange}
              >
                {option.label || <strong>Unlabelled option</strong>}
              </ToggleChip>
            ))}
        </MultipleChoiceAnswerOptions>
        {this.handleError()}
      </>
    );
  }

  renderCheckboxOptionSelector(hasError, hasConditionError) {
    const { expression } = this.props;

    return (
      <>
        <MultipleChoiceAnswerOptions
          data-test="options-selector"
          hasError={hasError}
        >
          <Label>Match</Label>
          <ConditionSelect
            onChange={this.handleConditionChange}
            defaultValue={expression.condition}
            hasError={hasConditionError}
            data-test="condition-dropdown"
          >
            <option value={answerConditions.ANYOF}>Any of</option>
            <option value={answerConditions.NOTANYOF}>Not any of</option>
            <option value={answerConditions.ALLOF}>All of</option>
            {enableOn(["enableCountCondition"]) && (
              <option value={answerConditions.COUNTOF}>Count of</option>
            )}
            <option value={answerConditions.UNANSWERED}>Unanswered</option>
          </ConditionSelect>
          {expression.condition === answerConditions.COUNTOF &&
            enableOn(["enableCountCondition"]) && (
              <SecondaryConditionSelector
                expression={expression}
                onRightChange={this.props.onRightChange}
                onConditionChange={this.props.onConditionChange}
              />
            )}
          {expression.condition !== answerConditions.UNANSWERED &&
            expression.condition !== answerConditions.COUNTOF && (
              <>
                <TransitionGroup component={SelectedOptions}>
                  {get(expression, "right.options", []).map((option) => {
                    const isMutuallyExclusive =
                      expression.left.mutuallyExclusiveOption &&
                      option.id === expression.left.mutuallyExclusiveOption.id;

                    return (
                      <CheckboxChipTransition key={option.id}>
                        <CheckboxChip
                          key={option.id}
                          id={option.id}
                          onRemove={this.handleCheckboxUnselect}
                          isMutuallyExclusive={isMutuallyExclusive}
                        >
                          {option.label || <strong>Unlabelled option</strong>}
                        </CheckboxChip>
                      </CheckboxChipTransition>
                    );
                  })}
                </TransitionGroup>
                <ChooseButton
                  onClick={() => {
                    this.setState({
                      showPopup: true,
                    });
                  }}
                >
                  Choose
                </ChooseButton>
              </>
            )}
          {this.state.showPopup && (
            <Popover isOpen onClose={this.handlePickerClose}>
              <CheckboxOptionPicker
                expression={expression}
                onClose={this.handlePickerClose}
                onRightChange={this.props.onRightChange}
              />
            </Popover>
          )}
        </MultipleChoiceAnswerOptions>
        {this.handleError()}
      </>
    );
  }

  render() {
    const { expression, groupErrorMessage } = this.props;
    const answerType = get(expression, "left.type");

    const { errors } = expression.validationErrorInfo;

    const hasError = errors.length || groupErrorMessage;
    const hasConditionError =
      errors.filter(({ field }) => field === "condition").length > 0;

    if (answerType === RADIO) {
      return this.renderRadioOptionSelector(hasError);
    } else {
      return this.renderCheckboxOptionSelector(hasError, hasConditionError);
    }
  }
}

export default MultipleChoiceAnswerOptionsSelector;
