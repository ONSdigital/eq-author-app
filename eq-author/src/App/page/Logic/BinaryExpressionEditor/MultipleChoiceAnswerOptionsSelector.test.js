import React from "react";
import { shallow } from "enzyme";
import { render, act, flushPromises } from "tests/utils/rtl";

import { rightSideErrors } from "constants/validationMessages";
import { colors } from "constants/theme";

import { CHECKBOX, RADIO } from "constants/answer-types";

import MultipleChoiceAnswerOptionsSelector from "./MultipleChoiceAnswerOptionsSelector";
import ToggleChip from "components/buttons/ToggleChip";
import { enableOn } from "utils/featureFlags";

describe("MultipleChoiceAnswerOptionsSelector", () => {
  let defaultProps;

  beforeEach(() => {
    defaultProps = {
      expression: {
        id: "Expression-id1",
        left: {
          id: "1",
          type: RADIO,
          options: [
            { label: "a", id: "1" },
            { label: "b", id: "2" },
            { label: "c", id: "3" },
          ],
        },
        right: null,
        validationErrorInfo: {
          totalCount: 0,
          id: "Mult-pass",
          errors: [],
        },
      },
      onRightChange: jest.fn(),
      onConditionChange: jest.fn(),
    };
  });

  it("should render consistently", () => {
    const wrapper = shallow(
      <MultipleChoiceAnswerOptionsSelector {...defaultProps} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("should call onRightChange with the added option when option is checked", () => {
    const wrapper = shallow(
      <MultipleChoiceAnswerOptionsSelector {...defaultProps} />
    );

    wrapper
      .find(ToggleChip)
      .first()
      .simulate("change", { name: "1", value: true });

    expect(defaultProps.onRightChange).toHaveBeenCalledWith({
      selectedOptions: ["1"],
    });
  });

  it("should call onRightChange with the option removed when it is unchecked", () => {
    defaultProps.expression.right = {
      options: [{ id: "1" }, { id: "2" }],
    };
    const wrapper = shallow(
      <MultipleChoiceAnswerOptionsSelector {...defaultProps} />
    );

    wrapper
      .find(ToggleChip)
      .first()
      .simulate("change", { name: "2", value: false });

    expect(defaultProps.onRightChange).toHaveBeenCalledWith({
      selectedOptions: ["1"],
    });
  });

  it("should render a list of previously selected options correctly", () => {
    defaultProps.expression.right = {
      options: [{ id: "1" }, { id: "2" }],
    };

    const wrapper = shallow(
      <MultipleChoiceAnswerOptionsSelector {...defaultProps} />
    );

    expect(wrapper.find(ToggleChip).first().prop("checked")).toBeTruthy();
    expect(wrapper.find(ToggleChip).at(1).prop("checked")).toBeTruthy();
    expect(wrapper.find(ToggleChip).last().prop("checked")).toBeFalsy();
  });

  it("should show error message when right side empty", async () => {
    defaultProps.expression.right = null;
    defaultProps.expression.validationErrorInfo.errors[0] = {
      errorCode: rightSideErrors.ERR_RIGHTSIDE_NO_VALUE.errorCode,
      field: "right",
      id: "expression-routing-1-right",
      type: "expressions",
    };

    const { getByText } = render(
      <MultipleChoiceAnswerOptionsSelector hasError {...defaultProps} />
    );

    await act(async () => {
      await flushPromises();
    });

    expect(
      getByText(rightSideErrors.ERR_RIGHTSIDE_NO_VALUE.optionsMessage)
    ).toHaveStyleRule("width", "100%");

    expect(
      getByText(rightSideErrors.ERR_RIGHTSIDE_NO_VALUE.optionsMessage)
    ).toBeTruthy();
  });

  it("should show error message when group condition is incompatible", async () => {
    defaultProps.expression.condition = "AllOf";
    defaultProps.expression.left.type = CHECKBOX;
    defaultProps.expression.validationErrorInfo.errors[0] = {
      errorCode:
        rightSideErrors.ERR_GROUP_MIXING_EXPRESSIONS_WITH_OR_STND_OPTIONS_IN_AND
          .errorCode,
      field: "groupCondition",
      id: "123-123-123",
      type: "expressions",
    };

    const { getByText } = render(
      <MultipleChoiceAnswerOptionsSelector hasError {...defaultProps} />
    );

    await act(async () => {
      await flushPromises();
    });

    expect(
      getByText(
        rightSideErrors.ERR_GROUP_MIXING_EXPRESSIONS_WITH_OR_STND_OPTIONS_IN_AND
          .message
      )
    ).toHaveStyleRule("width", "100%");

    expect(
      getByText(
        rightSideErrors.ERR_GROUP_MIXING_EXPRESSIONS_WITH_OR_STND_OPTIONS_IN_AND
          .message
      )
    ).toBeTruthy();
  });

  it("should show error message when exclusive OR error", async () => {
    defaultProps.expression.right = null;
    defaultProps.expression.validationErrorInfo.errors[0] = {
      errorCode: rightSideErrors.ERR_RIGHTSIDE_ALLOFF_OR_NOT_ALLOWED.errorCode,
      field: "right",
      id: "expression-routing-1-right",
      type: "expressions",
    };

    const { getByText } = render(
      <MultipleChoiceAnswerOptionsSelector hasError {...defaultProps} />
    );

    await act(async () => {
      await flushPromises();
    });

    expect(
      getByText(rightSideErrors.ERR_RIGHTSIDE_ALLOFF_OR_NOT_ALLOWED.message)
    ).toHaveStyleRule("width", "100%");

    expect(
      getByText(rightSideErrors.ERR_RIGHTSIDE_ALLOFF_OR_NOT_ALLOWED.message)
    ).toBeTruthy();
  });

  it("should display validation error when expression group-wide message passed in", async () => {
    const errorMessage = "Test group error message";
    defaultProps.groupErrorMessage = errorMessage;

    const { getByText } = render(
      <MultipleChoiceAnswerOptionsSelector hasError {...defaultProps} />
    );

    await act(async () => {
      await flushPromises();
    });

    expect(getByText(errorMessage)).toBeTruthy();

    expect(getByText(errorMessage)).toHaveStyleRule("width", "100%");
  });

  it("should highlight the condition selector when the error includes it", async () => {
    defaultProps.expression = {
      left: {},
      validationErrorInfo: {
        errors: [
          {
            errorCode:
              rightSideErrors.ERR_RIGHTSIDE_MIXING_OR_STND_OPTIONS_IN_AND_RULE
                .errorCode,
            field: "condition",
            id: "expression-routing-1-right",
            type: "expressions",
          },
        ],
      },
    };

    const { getByTestId } = render(
      <MultipleChoiceAnswerOptionsSelector hasError {...defaultProps} />
    );

    await act(async () => {
      await flushPromises();
    });

    const dropdown = getByTestId("condition-dropdown");

    expect(dropdown).toHaveStyle(`border: 2px solid ${colors.errorPrimary}`);
  });

  it("should include or in option label when mutually exclusive", async () => {
    defaultProps.expression.left = {
      ...defaultProps.expression.left,
      mutuallyExclusiveOption: { id: "123", label: "hello world" },
      type: CHECKBOX,
    };
    defaultProps.expression.right = {
      options: [
        { label: "a", id: "1" },
        { label: "b", id: "2" },
        { label: "hello world", id: "123" },
      ],
    };

    const { getByText, getByTestId } = render(
      <MultipleChoiceAnswerOptionsSelector {...defaultProps} />
    );

    await act(async () => {
      await flushPromises();
    });

    expect(getByTestId("mutually-exclusive-separator")).toBeTruthy();
    expect(getByText("or")).toBeTruthy();
    expect(getByText("hello world")).toBeTruthy();
  });

  it("should show secondaryCondition selector and number input when when condition=`CountOf`", async () => {
    defaultProps.expression.left = {
      ...defaultProps.expression.left,
      mutuallyExclusiveOption: { id: "123", label: "hello world" },
      type: CHECKBOX,
    };
    defaultProps.expression.condition = "CountOf";

    const { getByTestId } = render(
      <MultipleChoiceAnswerOptionsSelector {...defaultProps} />
    );

    await act(async () => {
      await flushPromises();
    });
    expect(enableOn(["enableCountCondition"])).toBe(true);
    expect(getByTestId("secondaryCondition-selector")).toBeTruthy();
    expect(getByTestId("secondaryCondition-number-input")).toBeTruthy();
  });

  it("should show secondaryCondition selector error when condition=`CountOf` and selector=null", async () => {
    const errorMessage = "Choose an operator";

    defaultProps.expression.left = {
      ...defaultProps.expression.left,
      mutuallyExclusiveOption: { id: "123", label: "hello world" },
      type: CHECKBOX,
    };
    defaultProps.expression.condition = "CountOf";
    defaultProps.expression.secondaryCondition = null;
    defaultProps.expression.validationErrorInfo.errors[0] = {
      errorCode: "ERR_SEC_CONDITION_NOT_SELECTED",
      field: "secondaryCondition",
      id: "123-123-123",
      type: "routingExpression",
    };

    const { getByText } = render(
      <MultipleChoiceAnswerOptionsSelector hasError {...defaultProps} />
    );

    await act(async () => {
      await flushPromises();
    });
    expect(getByText(errorMessage)).toBeTruthy();
  });
});
