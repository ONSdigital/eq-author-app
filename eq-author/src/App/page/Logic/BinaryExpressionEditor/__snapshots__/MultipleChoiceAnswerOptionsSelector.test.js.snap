// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleChoiceAnswerOptionsSelector should render consistently 1`] = `
<Fragment>
  <MultipleChoiceAnswerOptionsSelector__MultipleChoiceAnswerOptions
    data-test="options-selector"
  >
    <MultipleChoiceAnswerOptionsSelector__Label>
      Match
    </MultipleChoiceAnswerOptionsSelector__Label>
    <MultipleChoiceAnswerOptionsSelector__ConditionSelect
      data-test="condition-dropdown"
      onChange={[Function]}
    >
      <option
        value="OneOf"
      >
        One of
      </option>
      <option
        value="Unanswered"
      >
        Unanswered
      </option>
    </MultipleChoiceAnswerOptionsSelector__ConditionSelect>
    <withChangeHandler(ToggleChip)
      checked={false}
      key="1"
      name="1"
      onChange={[Function]}
      title="a"
    >
      a
    </withChangeHandler(ToggleChip)>
    <withChangeHandler(ToggleChip)
      checked={false}
      key="2"
      name="2"
      onChange={[Function]}
      title="b"
    >
      b
    </withChangeHandler(ToggleChip)>
    <withChangeHandler(ToggleChip)
      checked={false}
      key="3"
      name="3"
      onChange={[Function]}
      title="c"
    >
      c
    </withChangeHandler(ToggleChip)>
  </MultipleChoiceAnswerOptionsSelector__MultipleChoiceAnswerOptions>
</Fragment>
`;
